tools/evict-shared-cache/evict-shared-cache.cpp

```diff
@@ -0,0 +1,188 @@
+/*
+ * Copyright 2023 SiFive, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You should have received a copy of LICENSE.Apache2 along with
+ * this software. If not, you may obtain a copy at
+ *
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Open Group Base Specifications Issue 7
+#define _XOPEN_SOURCE 700
+#define _POSIX_C_SOURCE 200809L
+
+#include <fcntl.h>
+#include <string.h>
+#include <unistd.h>
+
+#include <algorithm>
+#include <iostream>
+#include <memory>
+#include <vector>
+
+#include "command.h"
+#include "eviction-policy.h"
+#include "gopt/gopt-arg.h"
+#include "gopt/gopt.h"
+
+void print_help(const char* argv0) {
+  // clang-format off
+  std::cerr << std::endl
+     << "Usage: " << argv0 << " [OPTIONS]" << std::endl
+     << "  --cache  DIR     Evict from shared cache DIR"  << std::endl
+     << "  --policy POLICY  Evict using POLICY"           << std::endl
+     << "  --help   -h      Print this message and exit"  << std::endl
+     << "Commands (read from stdin):" << std::endl
+     << "  write JOB_ID     JOB_ID was written into the shared cache" << std::endl
+     << "  read JOB_ID      JOB_ID was read from the shared cache" << std::endl
+     << "Available Policies:" << std::endl
+     << "  nil              No op policy. Process commands but do nothing." << std::endl
+     << std::endl;
+  // clang-format on
+}
+
+std::unique_ptr<EvictionPolicy> make_policy(const char* argv0, const char* policy) {
+  if (strcmp(policy, "nil") == 0) {
+    return std::make_unique<NilEvictionPolicy>();
+  }
+
+  std::cerr << "Unknown policy: " << policy << std::endl;
+  print_help(argv0);
+  exit(EXIT_FAILURE);
+}
+
+bool stdstreams_still_open() {
+  int status = fcntl(STDIN_FILENO, F_GETFD);
+
+  if (status == -1 && errno == EBADF) {
+    return false;
+  }
+
+  status = fcntl(STDOUT_FILENO, F_GETFD);
+
+  if (status == -1 && errno == EBADF) {
+    return false;
+  }
+
+  return true;
+}
+
+struct CommandParser {
+  std::string command_buff = "";
+
+  std::vector<std::string> read_commands() {
```

Comment:
Yeah I think I've concluded that we should use system polling here actually. Look at poll.h in src/runtime. You'll probably want to put that in src/util or src/wcl instead of in runtime so that its more widely usable.